File: src/main/java/com/kamikazejam/datastore/base/index/IndexedField.java
package com.kamikazejam.datastore.base.index;

import com.kamikazejam.datastore.base.Cache;
import com.kamikazejam.datastore.base.Store;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * All IndexFields are assumed to be unique (only have one Store with that value)
 */
@Getter @SuppressWarnings("unused")
public abstract class IndexedField<X extends Store<X, ?>, T> {
    private final @NotNull Cache<?, X> cache;
    private final @NotNull String name;
    public IndexedField(@NotNull Cache<?, X> cache, @NotNull String name) {
        this.cache = cache;
        this.name = name;
    }

    public abstract boolean equals(@Nullable T a, @Nullable T b);

    public abstract <K, Y extends Store<Y, K>> T getValue(Y store);

    public abstract @NotNull String toString(@NotNull Object value);

    public abstract @NotNull T fromString(@NotNull String value);
}

File: src/main/kotlin/com/kamikazejam/datastore/base/index/IndexedField.kt
package com.kamikazejam.datastore.base.index

import com.kamikazejam.datastore.base.Collection
import com.kamikazejam.datastore.store.Store

/**
 * All IndexFields are assumed to be unique (only have one Store with that value)
 */
@Suppress("unused")
abstract class IndexedField<X : Store<X, *>, T>(
    private val collection: Collection<*, X>,
    val name: String
) {
    abstract fun equals(a: T?, b: T?): Boolean

    abstract fun getValue(store: X): T?
}
